File: aide/flow/hooks.py
"""环节钩子：PlantUML 与 CHANGELOG 校验。"""

from __future__ import annotations

import shutil
import subprocess
from pathlib import Path
from typing import Any

from aide.core import output
from aide.flow.errors import FlowError
from aide.flow.git import GitIntegration
from aide.flow.types import FlowStatus


def run_pre_commit_hooks(
    *,
    root: Path,
    git: GitIntegration,
    status: FlowStatus | None,
    from_phase: str | None,
    to_phase: str,
    action: str,
    config: dict[str, Any] | None = None,
) -> None:
    if from_phase == "flow-design" and action in {"next-part", "back-part"}:
        _hook_plantuml(root=root, config=config)
    if from_phase == "docs" and action in {"next-part", "back-part"}:
        _hook_changelog_on_leave_docs(root=root, git=git, status=status)
    if to_phase == "finish" and action == "next-part":
        _hook_clean_task_plans(root=root, config=config)


def run_post_commit_hooks(*, to_phase: str, action: str) -> None:
    if to_phase == "docs" and action in {"start", "next-part", "back-part"}:
        output.info("请更新 CHANGELOG.md")


def _get_plantuml_command(config: dict[str, Any] | None) -> list[str] | None:
    """获取 PlantUML 命令，优先使用配置的 jar 文件。"""
    if config:
        jar_path = config.get("plantuml", {}).get("jar_path")
        java_path = config.get("plantuml", {}).get("java_path", "java")

        if jar_path:
            # 尝试解析 jar 路径
            jar_file = Path(jar_path)
            if not jar_file.is_absolute():
                # 相对路径，相对于 aide-program 目录
                aide_program_dir = Path(__file__).parent.parent.parent
                jar_file = aide_program_dir / jar_path

            if jar_file.exists():
                return [java_path, "-jar", str(jar_file)]

    # 回退到系统 plantuml 命令
    if shutil.which("plantuml"):
        return ["plantuml"]

    return None


def _hook_plantuml(*, root: Path, config: dict[str, Any] | None = None) -> None:
    """PlantUML 校验和构建钩子。"""
    # 获取流程图目录
    diagram_path = ".aide/diagrams"
    if config:
        diagram_path = config.get("flow", {}).get("diagram_path", diagram_path)

    diagram_dir = root / diagram_path

    # 收集所有 .puml 文件
    candidates: list[Path] = []

    # 从配置的流程图目录
    if diagram_dir.exists():
        candidates.extend([p for p in diagram_dir.rglob("*.puml") if p.is_file()])
        candidates.extend([p for p in diagram_dir.rglob("*.plantuml") if p.is_file()])

    # 也检查 docs 和 discuss 目录（向后兼容）
    for base in (root / "docs", root / "discuss"):
        if not base.exists():
            continue
        candidates.extend([p for p in base.rglob("*.puml") if p.is_file()])
        candidates.extend([p for p in base.rglob("*.plantuml") if p.is_file()])

    if not candidates:
        return

    # 获取 PlantUML 命令
    plantuml_cmd = _get_plantuml_command(config)
    if plantuml_cmd is None:
        output.warn("未找到 PlantUML（jar 或系统命令），已跳过校验/PNG 生成")
        return

    # 先校验所有文件
    errors: list[str] = []
    for file_path in candidates:
        result = subprocess.run(
            plantuml_cmd + ["-checkonly", str(file_path)],
            cwd=root,
            text=True,
            capture_output=True,
        )
        if result.returncode != 0:
            detail = (result.stderr or "").strip() or (result.stdout or "").strip()
            errors.append(f"{file_path.name}: {detail}")

    if errors:
        error_msg = "\n".join(errors)
        raise FlowError(f"PlantUML 语法校验失败:\n{error_msg}")

    # 校验通过，生成 PNG
    for file_path in candidates:
        result = subprocess.run(
            plantuml_cmd + ["-tpng", str(file_path)],
            cwd=root,
            text=True,
            capture_output=True,
        )
        if result.returncode != 0:
            detail = (result.stderr or "").strip() or (result.stdout or "").strip()
            raise FlowError(f"PlantUML PNG 生成失败: {file_path} {detail}".strip())

    output.ok(f"PlantUML 处理完成: {len(candidates)} 个文件")


def _hook_changelog_on_leave_docs(*, root: Path, git: GitIntegration, status: FlowStatus | None) -> None:
    changelog = root / "CHANGELOG.md"
    if not changelog.exists():
        raise FlowError("离开 docs 前需要更新 CHANGELOG.md（当前文件不存在）")

    git.ensure_repo()
    if git.status_porcelain("CHANGELOG.md").strip():
        return

    if status is None:
        raise FlowError("离开 docs 前需要更新 CHANGELOG.md（未找到流程状态）")

    for entry in status.history:
        if entry.phase != "docs":
            continue
        if not entry.git_commit:
            continue
        if git.commit_touches_path(entry.git_commit, "CHANGELOG.md"):
            return

    raise FlowError("离开 docs 前需要更新 CHANGELOG.md（未检测到 docs 阶段的更新记录）")


def _hook_clean_task_plans(*, root: Path, config: dict[str, Any] | None) -> None:
    """进入 finish 时清理任务计划文件。"""
    # 获取任务计划目录配置
    plans_path = ".aide/task-plans"
    if config:
        plans_path = config.get("task", {}).get("plans_path", plans_path)

    # 移除末尾斜杠（如有）
    plans_path = plans_path.rstrip("/")

    plans_dir = root / plans_path
    if not plans_dir.exists():
        return

    # 收集所有文件
    files = [f for f in plans_dir.iterdir() if f.is_file()]
    if not files:
        return

    # 删除文件
    for f in files:
        f.unlink()

    output.ok(f"已清理任务计划文件: {len(files)} 个")

File: src/flow/hooks.rs
use std::fs;
use std::path::Path;
use std::process::Command;

use crate::core::output;
use crate::flow::git::GitIntegration;
use crate::flow::types::FlowStatus;

pub fn run_pre_commit_hooks(
    root: &Path,
    git: &GitIntegration,
    status: Option<&FlowStatus>,
    from_phase: Option<&str>,
    to_phase: &str,
    action: &str,
    config: &toml::Value,
) -> Result<(), String> {
    if from_phase == Some("flow-design") && (action == "next-part" || action == "back-part") {
        hook_plantuml(root, config)?;
    }
    if from_phase == Some("docs") && (action == "next-part" || action == "back-part") {
        hook_changelog_on_leave_docs(root, git, status)?;
    }
    if to_phase == "finish" && action == "next-part" {
        hook_clean_task_plans(root, config);
    }
    Ok(())
}

pub fn run_post_commit_hooks(to_phase: &str, action: &str) {
    if to_phase == "docs" && (action == "start" || action == "next-part" || action == "back-part") {
        output::info("请更新 CHANGELOG.md");
    }
}

fn get_plantuml_command(config: &toml::Value) -> Option<Vec<String>> {
    let jar_path = config
        .get("plantuml")
        .and_then(|p| p.get("jar_path"))
        .and_then(|v| v.as_str())
        .unwrap_or("");

    if !jar_path.is_empty() {
        let jar_file = std::path::PathBuf::from(jar_path);
        let jar_file = if jar_file.is_absolute() {
            jar_file
        } else {
            // 相对路径：相对于可执行文件目录
            let exe_dir = std::env::current_exe()
                .ok()
                .and_then(|e| e.parent().map(|p| p.to_path_buf()))
                .unwrap_or_else(|| std::path::PathBuf::from("."));
            exe_dir.join(jar_path)
        };

        if jar_file.exists() {
            let java_path = config
                .get("plantuml")
                .and_then(|p| p.get("java_path"))
                .and_then(|v| v.as_str())
                .unwrap_or("java");
            return Some(vec![
                java_path.to_string(),
                "-jar".to_string(),
                jar_file.to_string_lossy().to_string(),
            ]);
        }
    }

    // 回退到系统 plantuml 命令
    if Command::new("plantuml")
        .arg("--version")
        .output()
        .is_ok_and(|o| o.status.success())
    {
        return Some(vec!["plantuml".to_string()]);
    }

    None
}

fn hook_plantuml(root: &Path, config: &toml::Value) -> Result<(), String> {
    let diagram_path = config
        .get("flow")
        .and_then(|f| f.get("diagram_path"))
        .and_then(|v| v.as_str())
        .unwrap_or(".aide/diagrams");

    let diagram_dir = root.join(diagram_path);

    let mut candidates: Vec<std::path::PathBuf> = Vec::new();

    // 收集 .puml / .plantuml 文件
    for dir in [&diagram_dir, &root.join("docs"), &root.join("discuss")] {
        if dir.exists() {
            collect_puml_files(dir, &mut candidates);
        }
    }

    if candidates.is_empty() {
        return Ok(());
    }

    let plantuml_cmd = match get_plantuml_command(config) {
        Some(cmd) => cmd,
        None => {
            output::warn("未找到 PlantUML（jar 或系统命令），已跳过校验/PNG 生成");
            return Ok(());
        }
    };

    // 语法检查
    let mut errors = Vec::new();
    for file_path in &candidates {
        let mut cmd = Command::new(&plantuml_cmd[0]);
        for arg in &plantuml_cmd[1..] {
            cmd.arg(arg);
        }
        cmd.arg("-checkonly").arg(file_path).current_dir(root);

        let result = cmd.output();
        if let Ok(output) = result {
            if !output.status.success() {
                let detail = String::from_utf8_lossy(&output.stderr).trim().to_string();
                let detail = if detail.is_empty() {
                    String::from_utf8_lossy(&output.stdout).trim().to_string()
                } else {
                    detail
                };
                errors.push(format!(
                    "{}: {detail}",
                    file_path.file_name().unwrap_or_default().to_string_lossy()
                ));
            }
        }
    }

    if !errors.is_empty() {
        return Err(format!("PlantUML 语法校验失败:\n{}", errors.join("\n")));
    }

    // 生成 PNG
    for file_path in &candidates {
        let mut cmd = Command::new(&plantuml_cmd[0]);
        for arg in &plantuml_cmd[1..] {
            cmd.arg(arg);
        }
        cmd.arg("-tpng").arg(file_path).current_dir(root);

        let result = cmd.output();
        if let Ok(output) = result {
            if !output.status.success() {
                let detail = String::from_utf8_lossy(&output.stderr).trim().to_string();
                return Err(format!(
                    "PlantUML PNG 生成失败: {} {detail}",
                    file_path.display()
                ));
            }
        }
    }

    output::ok(&format!("PlantUML 处理完成: {} 个文件", candidates.len()));
    Ok(())
}

fn collect_puml_files(dir: &Path, candidates: &mut Vec<std::path::PathBuf>) {
    if let Ok(entries) = fs::read_dir(dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_dir() {
                collect_puml_files(&path, candidates);
            } else if path.is_file() {
                if let Some(ext) = path.extension() {
                    let ext = ext.to_string_lossy();
                    if ext == "puml" || ext == "plantuml" {
                        candidates.push(path);
                    }
                }
            }
        }
    }
}

fn hook_changelog_on_leave_docs(
    root: &Path,
    git: &GitIntegration,
    status: Option<&FlowStatus>,
) -> Result<(), String> {
    let changelog = root.join("CHANGELOG.md");
    if !changelog.exists() {
        return Err("离开 docs 前需要更新 CHANGELOG.md（当前文件不存在）".into());
    }

    git.ensure_repo()?;

    // 检查工作目录中是否有未暂存的 CHANGELOG.md 变更
    let porcelain = git.status_porcelain("CHANGELOG.md")?;
    if !porcelain.trim().is_empty() {
        return Ok(());
    }

    let status = match status {
        Some(s) => s,
        None => {
            return Err("离开 docs 前需要更新 CHANGELOG.md（未找到流程状态）".into());
        }
    };

    for entry in &status.history {
        if entry.phase != "docs" {
            continue;
        }
        if let Some(commit) = &entry.git_commit {
            if git.commit_touches_path(commit, "CHANGELOG.md")? {
                return Ok(());
            }
        }
    }

    Err("离开 docs 前需要更新 CHANGELOG.md（未检测到 docs 阶段的更新记录）".into())
}

fn hook_clean_task_plans(root: &Path, config: &toml::Value) {
    let plans_path = config
        .get("task")
        .and_then(|t| t.get("plans_path"))
        .and_then(|v| v.as_str())
        .unwrap_or(".aide/task-plans")
        .trim_end_matches('/');

    let plans_dir = root.join(plans_path);
    if !plans_dir.exists() {
        return;
    }

    let mut count = 0;
    if let Ok(entries) = fs::read_dir(&plans_dir) {
        for entry in entries.flatten() {
            let path = entry.path();
            if path.is_file() {
                let _ = fs::remove_file(&path);
                count += 1;
            }
        }
    }

    if count > 0 {
        output::ok(&format!("已清理任务计划文件: {count} 个"));
    }
}
